[DOC] New documentation: Self-host maps server (#1625)
opensearch-project/documentation-website: _dashboards/selfhost-maps-server.md

---
layout: default
title: Self-host maps server
nav_order: 6
---

# Self-host maps server

The self-host maps server for OpenSearch Dashboards allows users to access the default maps service on air-gapped environments. The server provides a map manifest with map tiles and vectors URL, map tiles URL, and map vectors URL that are compatible with OpenSearch.

> The following steps walk through setting up and using the self-host maps server with OpenSearch Dashboards.

Reviewer: ```suggestion
The following sections provide steps for setting up and using the self-host maps server with OpenSearch Dashboards.
```

> You can access the `maps-server` image via the OpenSearch official [Docker Hub repository](https://hub.docker.com/u/opensearchproject).

Reviewer: ```suggestion
You can access the `maps-server` image via the official OpenSearch [Docker Hub repository](https://hub.docker.com/u/opensearchproject).
```

{: .note}

## Pull Docker image

Open your terminal and run the following command:

`docker pull opensearch/opensearch-maps-server`

## Set up the server

> You must set up the map tiles before running the server. You have two setup options: 

Reviewer: We can't introduce headings with a colon. Instead, name the two options after the colon.

> ### Option 1: Use the OpenSearch-provided maps service tiles set

Reviewer: Here this is "tiles set", but we use "tile set" below. Let's make consistent, as appropriate.

Create a Docker volume to hold the tiles set:

`docker volume create tiles-data`

> Download the tiles set from the OpenSearch maps service. Planet tiles sets are available for up to 8 zoom levels `https://maps.opensearch.org/offline/planet-osm-default-z0-z8.tar.gz` and up to 10 zoom levels `https://maps.opensearch.org/offline/planet-osm-default-z0-z10.tar.gz`.

Reviewer: Should the links be in parentheses or preceded by "at"?

Author: Revised as follows: Two planet tiles sets are available based on the desired zoom level:

- Zoom Level 8 (https://maps.opensearch.org/offline/planet-osm-default-z0-z8.tar.gz)
- Zoom level 10 (https://maps.opensearch.org/offline/planet-osm-default-z0-z10.tar.gz)

The planet tiles set for zoom level 10 (2 GB compressed/6.8 GB uncompressed) is approximately 10 times larger than the set for zoom level 8 (225 MB compressed/519 MB uncompressed).

```
docker run \
    -e DOWNLOAD_TILES=https://maps.opensearch.org/offline/planet-osm-default-z0-z8.tar.gz \
    -v tiles-data:/usr/src/app/public/tiles/data/ \
    opensearch/opensearch-maps-server \
    import
```

### Option 2: Generate the raster tile set

> To generate the raster tiles images set, use the [raster tile generation pipeline](https://github.com/opensearch-project/maps/tree/main/tiles-generation/cdk) and then use the tile set absolute path to create a volume to start the server.

Reviewer: "raster tiles images" seems to be inconsistent with "raster tile" and "tile set". Let's make consistent, as appropriate.

## Start the server

> Use the following command to start the server using the Docker volume `tiles-data`. The command below is an example using host URL "localhost" and port "8080".  

Reviewer: ```suggestion
Use the following command to start the server using the Docker volume `tiles-data`. The following command is an example using host URL "localhost" and port "8080".  
```

```
docker run \
    -v tiles-data:/usr/src/app/public/tiles/data/ \
    -e HOST_URL='http://localhost' \
    -p 8080:8080 \
    opensearch/opensearch-maps-server \
    run
```

> Or, if you generated raster tiles images, run the server using that tiles set. 

Reviewer: Same comment re: consistency of tile/tiles.

```
docker run \
    -v /absolute/path/to/tiles/:/usr/src/app/dist/public/tiles/data/ \
    -p 8080:8080 \
    opensearch/opensearch-maps-server \
    run
```

> To access the images, open those URLs in a browser on the host or use curl `curl http://localhost:8080/manifest.json`. 

Reviewer: Should "curl" be "cURL"? Should a colon follow "cURL"?

> To confirm the server is running, you can access the following files on your localhost. To access the file, open the URLs in a browser on the host or use `curl http://localhost:8080/manifest.json`.

Reviewer: ```suggestion
To confirm the server is running, you can access the following files on your local host. To access the file, open the URLs in a browser on the host or use `curl http://localhost:8080/manifest.json`.
```

---
layout: default
title: Self-host maps server
nav_order: 6
---

# Self-host maps server

The self-host maps server for OpenSearch Dashboards allows users to access the default maps service in air-gapped environments. OpenSearch-compatible map URLs include a map manifest with map tiles and vectors, the map tiles, and the map vectors.

The following sections provide steps for setting up and using the self-host maps server with OpenSearch Dashboards.

You can access the `maps-server` image via the official OpenSearch [Docker Hub repository](https://hub.docker.com/u/opensearchproject).
{: .note}

## Pull Docker image

Open your terminal and run the following command:

`docker pull opensearch/opensearch-maps-server`

## Set up the server

> You must set up the map tiles before running the server. You have two setup options: use the OpenSearch-provided maps service tiles set or generate the raster tiles set.

Reviewer: ```suggestion
You must set up the map tiles before running the server. You have two setup options: Use the OpenSearch-provided maps service tiles set, or generate the raster tiles set.
```

### Option 1: Use the OpenSearch-provided maps service tiles set

Create a Docker volume to hold the tiles set:

`docker volume create tiles-data`

Download the tiles set from the OpenSearch maps service. Two planet tiles sets are available based on the desired zoom level:

- Zoom Level 8 (https://maps.opensearch.org/offline/planet-osm-default-z0-z8.tar.gz)
- Zoom level 10 (https://maps.opensearch.org/offline/planet-osm-default-z0-z10.tar.gz)

The planet tiles set for zoom level 10 (2 GB compressed/6.8 GB uncompressed) is approximately 10 times larger than the set for zoom level 8 (225 MB compressed/519 MB uncompressed).
{: .note} 

```
docker run \
    -e DOWNLOAD_TILES=https://maps.opensearch.org/offline/planet-osm-default-z0-z8.tar.gz \
    -v tiles-data:/usr/src/app/public/tiles/data/ \
    opensearch/opensearch-maps-server \
    import
```

### Option 2: Generate the raster tiles set

To generate the raster tiles set, use the [raster tile generation pipeline](https://github.com/opensearch-project/maps/tree/main/tiles-generation/cdk) and then use the tiles set absolute path to create a volume to start the server.

## Start the server

> Use the following command to start the server using the Docker volume `tiles-data`. The following command is an example using host URL "localhost" and port "8080".

Reviewer: ```suggestion
Use the following command to start the server using the Docker volume `tiles-data`. The following command is an example using host URL "localhost" and port "8080":
```

```
docker run \
    -v tiles-data:/usr/src/app/public/tiles/data/ \
    -e HOST_URL='http://localhost' \
    -p 8080:8080 \
    opensearch/opensearch-maps-server \
    run
```

> Or, if you generated the raster tiles set, run the server using that tiles set. 

Reviewer: ```suggestion
Or, if you generated the raster tiles set, run the server using that tiles set:
```

---
layout: default
title: Self-host maps server
nav_order: 6
---

# Self-host maps server

> The self-host maps server for OpenSearch Dashboards allows users to access the default maps service on air-gapped environments. The server provides a map manifest url with map tiles and vectors, a map tiles url, and a map vectors url that are compatible with OpenSearch.

Reviewer 2: AWS appears to capitalize URL

Reviewer 3: Yeah - URL should be caps both instances

Author: revised

The following steps walk through setting up and using the self-host maps server with OpenSearch Dashboards.

You can access the `maps-server` image via the OpenSearch official [Docker Hub repository](https://hub.docker.com/u/opensearchproject).
{: .note}

## Pull Docker image

> Open Docker and then open your terminal to run the following command:

Reviewer 3: The "open docker" step only applies to people using Docker Desktop. You could remove "Open Docker" or you could replace it with "After installing Docker, make sure that the Docker daemon is running. Then, open your terminal [...]"

`docker pull opensearch/opensearch-maps-server`

## Set up the server

> You must set up the map tiles before running the server. You have two setup options: (1) use the OpenSearch-provided maps service tiles set or (2) generate the raster tile set yourself.

Reviewer 3: Break it out into a numbered list:

You have two setup options:
1. Option 1
2. Option 2

It looks like your bullet points already match the headings for those sections too so good work!!

### Option 1: Use the OpenSearch-provided maps service tiles set

Create a Docker volume to hold the tiles set:

`docker volume create tiles-data`

> Download the tiles set from the OpenSearch maps service. Planet tiles sets 0-8 `https://maps.opensearch.org/offline/planet-osm-default-z0-z8.tar.gz` and 0-10 `https://maps.opensearch.org/offline/planet-osm-default-z0-z10.tar.gz` are available.

Reviewer 3: Why would a user choose one over the other?

Reviewer 4: zoom range 0-8 tiles set is light download and use, 0-10 tiles set is heavier to download and use.

> Download the tiles set from the OpenSearch maps service (planet tiles sets [0-8](https://maps.opensearch.org/offline/planet-osm-default-z0-z8.tar.gz) and [0-10](https://maps.opensearch.org/offline/planet-osm-default-z0-z10.tar.gz) are available):

Reviewer 4: IMO, we can explicitly show the 0-8 and 0-10 url on doc so that user can be easier to use it in command. Thanks!
https://maps.opensearch.org/offline/planet-osm-default-z0-z8.tar.gz
https://maps.opensearch.org/offline/planet-osm-default-z0-z10.tar.gz

```
docker run \

> -e DOWNLOAD_TILES=https://maps.opensearch.org/offline/planet-osm-default-z0-z8.tar.gz \

Reviewer 3: indent all the lines for this command after the first line. So like:

```bash
docker run \
    -e DOWNLOAD_TILES=https: [...]
```

-v tiles-data:/usr/src/app/public/tiles/data/ \
opensearch/opensearch-maps-server \
import
```

### Option 2: Generate the raster tile set

To generate the raster tiles images set, use the [raster tile generation pipeline](https://github.com/opensearch-project/maps/tree/main/tiles-generation/cdk) and then use the tile set absolute path to create a volume to start the server.

## Start the server

> Use the following command to start the server using the Docker volume `tiles-data`. The command below is an example using host url "localhost" and port "8080."  

Reviewer 4: ```suggestion
Use the following command to start the server using the Docker volume `tiles-data`. The command below is an example using host url "localhost" and port "8080".  
```

```
docker run \

> -v tiles-data:/usr/src/app/public/tiles/data/ \

Reviewer 3: Add tab (or 4 spaces) to lines 52 through 56 so the command is more clear.

-e HOST_URL='http://localhost' \
-p 8080:8080 \
opensearch/opensearch-maps-server \
run
```

Or, if you generated raster tiles images, run the server using that tiles set:

```
docker run \

> -v /absolute/path/to/tiles/:/usr/src/app/dist/public/tiles/data/ \

Reviewer 3: Tab / add 4 spaces to lines 63 to 66

-p 8080:8080 \
opensearch/opensearch-maps-server \
run
```

> To confirm the server is running, you should be able to access the following files on your localhost:  

Reviewer 3: I would add steps here for how you would access them. You could do one of two things - open those URLs in a browser on the host, or use `curl` like this:

```bash
curl http://localhost:8080/manifest.json
```

Reviewer 3: Either way, the output will be a JSON file (whatever is in that manifest)

* [Maps manifest](http://localhost:8080/manifest.json)

> * [Maps tiles](http://localhost:8080/tiles/data/%7Bz%7D/%7Bx%7D/%7By%7D.png)

Author: revised

> * [Maps tiles demo](http://localhost:8080/) 

Reviewer 4: Would prefer to explicitly show the url, because it's a demo url using localhost, user most likely use public ip address instead.

Author: revised

## Use the self-host maps server with OpenSearch Dashboards

> You can use the self-host maps server with OpenSearch Dashboards by either adding the parameter to `opensearch_dashboards.yml` or configuring the default WMS properties in OpenSearch Dashboards.

Reviewer 3: Consider using a bullet-point list here like the other process outline at the top of the doc.

You can use the self-host maps server with OpenSearch Dashboards by either of the following methods:
1. Add the parameter...
2. Configure the default WMS...

Reviewer 3: Sorry, I used numbers instead of bullets. I think maybe bullets work better since they are options and not ordered steps? Hm. I'm on the fence there.

### Option 1: Configure opensearch_dashboards.yml

> Configure the manifest url in `opensearch_dashboards.yml`:

Reviewer 3: Capitalize URL

`map.opensearchManifestServiceUrl: "http://localhost:8080/manifest.json"`

### Option 2: Configure Default WMS properties in OpenSearch Dashboards

1. On the OpenSearch Dashboards console, select **Stack Management > Advanced Settings**. 
2. Locate `visualization:tileMap:WMSdefaults` under **Default WMS properties**. 

> 3. Change enabled to `true` and add the URL for the valid map server.

Reviewer 3: Since "enabled" is an actual setting in the file, maybe use it in monospace font like `enabled` - that way it's easy for the reader to tell the difference between non-inclusive language and literal settings in the application

## Licenses

Tiles are generated per [Terms of Use for Natural Earth vector map data](https://www.naturalearthdata.com/about/terms-of-use/) and [Copyright and License for OpenStreetMap](https://www.openstreetmap.org/copyright).

> ## Related links

Reviewer 2: I think Jeff might point this out. Do we need the link prefix changed for these?

{{site.url}}{{site.baseurl}}/

Reviewer 3: Good call out Chris!  But those aren't needed here since they're external links.

> Tiles are generated per [Terms of Use for Natural Earth raster and vector map data](https://www.naturalearthdata.com/about/terms-of-use/) and [Copyright and License for OpenStreetMap](https://www.openstreetmap.org/copyright).

Reviewer 4: From Natural Earth, We only vector map data from them. So we could say:

Tiles are generated per [Terms of Use for Natural Earth](https://www.naturalearthdata.com/about/terms-of-use/) vector map data and [Copyright and License for OpenStreetMap](https://www.openstreetmap.org/copyright). 

Thanks!
